### ISSUE TITLE
Parsing errors for quoted "MixedCase" custom data types with postgres

### ISSUE THREAD

**Issue by jeffmhastings** [sqlfluff/sqlfluff#4125]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Parsing failed for a create table containing a quoted custom data type.

### Expected Behaviour

Parsing shouldn't fail when custom data types are quoted.

### Observed Behaviour

Parsing fails with an "unparsable section" error.

```sh
$ sqlfluff lint --dialect postgres /tmp/script.sql
== [/tmp/script.sql] FAIL
L:   2 | P:   1 |  PRS | Line 2, Position 1: Found unparsable section: 'CREATE
                       | TABLE "User" (\n "id" SERIAL,\n ...'
WARNING: Parsing errors found and dialect is set to 'postgres'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

The problem has to do with the quoted custom type `"Mood"` in the definition of the mood column. When postgres identifiers contain capital letters, they must be double quoted since otherwise they will be treated as all lowercase.

Changing `"Mood"` to `Mood` or `mood`, for instance, resolves the problem.

This seems to be related to the use of `DataTypeIdentifierSegment` to support custom data types in the postgres dialect.

See [here](https://github.com/sqlfluff/sqlfluff/blob/16c34a26daf1a39291d0fc0585097dd8d1b47c2d/src/sqlfluff/dialects/dialect_postgres.py#L687) and [here](https://github.com/sqlfluff/sqlfluff/blob/fe771937e71ce1dc1837be4ada89fc427ef3d7ab/src/sqlfluff/dialects/dialect_ansi.py#L356)

### How to reproduce

Using the following SQL script

```sql
CREATE TYPE "Mood" AS ENUM ('sad', 'ok', 'happy');
CREATE TABLE "User" (
    "id" SERIAL,
    "mood" "Mood" NOT NULL
    );
```

run

```sh
sqlfluff --dialect postgres script.sql
```

### Dialect

postgres

### Version

```sh
$ sqlfluff --version
sqlfluff, version 1.4.2
```

```sh
$ python --version
Python 3.8.13
```

### Configuration

n/a

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

I'm happy to submit a PR to fix, but need some help identifying a proper fix that aligns with project standards.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by keraion** [comment]
Fixed by #4747
